toolchain/lexer/tokenized_buffer.cpp

```diff
@@ -247,6 +248,10 @@ class TokenizedBuffer::Lexer {
     bool formed_token_;
   };
 
+  using DispatchFunctionT = LexResult(Lexer& lexer,
+                                      llvm::StringRef& source_text);
```

Comment:
```suggestion
  using DispatchFunctionT = auto (Lexer& lexer,
                                  llvm::StringRef& source_text) -> LexResult;
```
I think our preference for trailing return types should apply here too.